pkg/server/endpoints/entryfetcher_test.go

```diff
@@ -0,0 +1,199 @@
+package endpoints
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"net/url"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/pkg/server/plugin/datastore"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/spiffe/spire/proto/spire/types"
+	"github.com/spiffe/spire/test/fakes/fakedatastore"
+	"github.com/spiffe/spire/test/fakes/fakemetrics"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	spiffeScheme = "spiffe"
+	trustDomain  = "example.org"
+)
+
+func TestNewAuthorizedEntryFetcherWithFullCache(t *testing.T) {
+	ctx := context.Background()
+	log := logrus.New()
+	log.Out = ioutil.Discard
```

Comment:
Nice, I missed that we had this in our set of test utilities. For some reason verifying that the logs were emitted in `TestRunRebuildCacheTask` wasn't working for me. I tried it with this patch:
```diff
diff --git a/pkg/server/endpoints/entryfetcher_test.go b/pkg/server/endpoints/entryfetcher_test.go
index d84caed3..bc044acc 100644
--- a/pkg/server/endpoints/entryfetcher_test.go
+++ b/pkg/server/endpoints/entryfetcher_test.go
@@ -7,12 +7,14 @@ import (
        "testing"

        "github.com/andres-erbsen/clock"
+       "github.com/sirupsen/logrus"
        "github.com/sirupsen/logrus/hooks/test"
        "github.com/spiffe/go-spiffe/v2/spiffeid"
        "github.com/spiffe/spire/pkg/server/api"
        "github.com/spiffe/spire/pkg/server/cache/entrycache"
        "github.com/spiffe/spire/proto/spire/common"
        "github.com/spiffe/spire/proto/spire/types"
+       "github.com/spiffe/spire/test/spiretest"
        "github.com/stretchr/testify/assert"
        "github.com/stretchr/testify/require"
 )
@@ -91,7 +93,7 @@ func TestFetchRegistrationEntries(t *testing.T) {

 func TestRunRebuildCacheTask(t *testing.T) {
        ctx, cancel := context.WithCancel(context.Background())
-       log, _ := test.NewNullLogger()
+       log, hook := test.NewNullLogger()
        clk := clock.NewMock()
        agentID := trustDomain.NewID("/root")
        var expectedEntries []*types.Entry
@@ -139,6 +141,7 @@ func TestRunRebuildCacheTask(t *testing.T) {

        // When the rebuild task is able to complete successfully,
        // the cache should now contain the Agent's new authorized entries
+       rErr := returnErr
        returnErr = nil
        clk.Add(cacheReloadInterval)
        entries, err = ef.FetchAuthorizedEntries(ctx, agentID)
@@ -147,6 +150,23 @@ func TestRunRebuildCacheTask(t *testing.T) {

        cancel()
        assert.NoError(t, <-errChan)
+       spiretest.AssertLogs(t, hook.AllEntries(), []spiretest.LogEntry{
+               {
+                       Level:   logrus.ErrorLevel,
+                       Message: "Failed to reload entry cache",
+                       Data: logrus.Fields{
+                               logrus.ErrorKey: rErr.Error(),
+                       },
+               },
+               {
+                       Level:   logrus.DebugLevel,
+                       Message: "Reloaded entry cache",
+               },
+               {
+                       Level:   logrus.DebugLevel,
+                       Message: "Stopping in-memory entry cache hydrator",
+               },
+       })
 }

 func setupExpectedEntriesData(t *testing.T, agentID spiffeid.ID) []*types.Entry {
```

but I keep getting the following error:

```
--- FAIL: TestRunRebuildCacheTask (0.00s)
    logs.go:24:
        	Error Trace:	logs.go:24
        	            				entryfetcher_test.go:153
        	Error:      	Not equal:
        	            	expected: []spiretest.LogEntry{spiretest.LogEntry{Level:0x2, Message:"Failed to reload entry cache", Data:logrus.Fields{"error":"some cache build error"}}, spiretest.LogEntry{Level:0x5, Message:"Reloaded entry cache", Data:logrus.Fields(nil)}, spiretest.LogEntry{Level:0x5, Message:"Stopping in-memory entry cache hydrator", Data:logrus.Fields(nil)}}
        	            	actual  : []spiretest.LogEntry(nil)

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,20 +1,2 @@
        	            	-([]spiretest.LogEntry) (len=3) {
        	            	- (spiretest.LogEntry) {
        	            	-  Level: (logrus.Level) error,
        	            	-  Message: (string) (len=28) "Failed to reload entry cache",
        	            	-  Data: (logrus.Fields) (len=1) {
        	            	-   (string) (len=5) "error": (string) (len=22) "some cache build error"
        	            	-  }
        	            	- },
        	            	- (spiretest.LogEntry) {
        	            	-  Level: (logrus.Level) debug,
        	            	-  Message: (string) (len=20) "Reloaded entry cache",
        	            	-  Data: (logrus.Fields) <nil>
        	            	- },
        	            	- (spiretest.LogEntry) {
        	            	-  Level: (logrus.Level) debug,
        	            	-  Message: (string) (len=39) "Stopping in-memory entry cache hydrator",
        	            	-  Data: (logrus.Fields) <nil>
        	            	- }
        	            	-}
        	            	+([]spiretest.LogEntry) <nil>

        	Test:       	TestRunRebuildCacheTask
        	Messages:   	unexpected logs
FAIL
FAIL	github.com/spiffe/spire/pkg/server/endpoints	4.562s
```

This failure doesn't really make any sense to me, since clearly the code is getting executed properly by virtue of the other assertions in the test. Is the test logger provided by `test.NewNullLogger()` safe for use across multiple goroutines? I'm wondering if this has something to do with the code emitting the logs in a separate goroutine from where the assertions are handled.